In [2]:
from pandas import Series, DataFrame
import pandas as pd
import numpy as np

# Series

In [2]:
obj = Series([4,7,-5,3])
obj

0    4
1    7
2   -5
3    3
dtype: int64

In [3]:
obj.values

array([ 4,  7, -5,  3])

In [7]:
type(obj.values)

numpy.ndarray

In [8]:
obj.index

Int64Index([0, 1, 2, 3], dtype='int64')

In [5]:
type(obj.index)

pandas.core.index.Int64Index

In [12]:
obj2 = Series([4,7,-5,3], 
              index=['d', 'b', 'a', 'c'])
obj2

d    4
b    7
a   -5
c    3
dtype: int64

In [15]:
obj2['d']

4

인덱스가 부여되어도 0 기반 인덱스가 기본적으로 계속 적용된다.

In [17]:
obj2[0]

4

In [18]:
obj2['b']= 3
obj2

d    4
b    3
a   -5
c    3
dtype: int64

값 선택: 지정한 순서대로 표시

In [19]:
obj2[['c','a','d']]

c    3
a   -5
d    4
dtype: int64

In [11]:
# 값이 0보다 큰 항목만 선택
obj2[obj2>0]

d    4
b    7
c    3
dtype: int64

In [4]:
[1,2,3]*2

[1, 2, 3, 1, 2, 3]

In [12]:
obj2 * 2

d     8
b    14
a   -10
c     6
dtype: int64

In [13]:
np.exp(obj2)

d      54.598150
b    1096.633158
a       0.006738
c      20.085537
dtype: float64

인덱스 포함 확인

In [24]:
'b' in obj2

True

In [25]:
'e' in obj2

False

값의 존재 여부 확인

In [14]:
4 in obj2.values

True

#### 연습문제

학생 세 명의 점수를 각각 Series 형식으로 저장한다. 점수는 국어, 영어, 수학 세 가지 점수를 가질 수 있다.


1. 각 학생의 점수를 수학, 국어 두 과목만 표시하되, 순서를 지켜 표시하시오.

2. 각 학생의 점수 중 70점 이상인 과목의 점수만을 출력하시오.

In [5]:
s1 = Series([60, 80, 90], index=['국', '영', '수'])
s1[s1>=70]

영    80
수    90
dtype: int64

dict --> Series

In [8]:
sdata = {'Ohio' : 35000, 
         'Texas': 71000, 
         'Oregon': 16000, 
         'Utah': 5000}
obj3 = Series(sdata)
obj3

Ohio      35000
Oregon    16000
Texas     71000
Utah       5000
dtype: int64

Series --> dict

In [6]:
dict(obj3)

{'Ohio': 35000, 'Oregon': 16000, 'Texas': 71000, 'Utah': 5000}

In [29]:
obj3.index

Index([u'Ohio', u'Oregon', u'Texas', u'Utah'], dtype='object')

In [30]:
obj3.values

array([35000, 16000, 71000,  5000])

In [9]:
# 원래 데이터에는 값이 없는 인덱스를 지정하는 경우
states = ['California', 'Ohio', 'Oregon', 'Texas']
obj4 = Series(sdata, index=states)
obj4

California      NaN
Ohio          35000
Oregon        16000
Texas         71000
dtype: float64

In [9]:
pd.isnull(obj4)

California   NaN
dtype: float64

In [10]:
obj4[pd.notnull(obj4)]

Ohio      35000
Oregon    16000
Texas     71000
dtype: float64

In [37]:
obj4.isnull()

CA         True
Ohio      False
Oregon    False
Texas     False
dtype: bool

In [11]:
obj3

Ohio      35000
Oregon    16000
Texas     71000
Utah       5000
dtype: int64

In [12]:
obj4

California      NaN
Ohio          35000
Oregon        16000
Texas         71000
dtype: float64

서로 인덱스가 어긋나기도 하는 두 Series의 연산

In [13]:
obj3+obj4

California       NaN
Ohio           70000
Oregon         32000
Texas         142000
Utah             NaN
dtype: float64

#### 연습문제

obj3와 obj4의 원소를 더하고, 값이 있는 것만 선택해 결과를 저장하시오.

In [10]:
obj = obj3 + obj4
obj[obj.notnull()]

Ohio       70000
Oregon     32000
Texas     142000
dtype: float64

In [26]:
obj4.name = 'population'
obj4

California      NaN
Ohio          35000
Oregon        16000
Texas         71000
Name: population, dtype: float64

In [29]:
obj4.name is 'population'

True

In [28]:
obj4.index.name = 'state'
obj4

state
California      NaN
Ohio          35000
Oregon        16000
Texas         71000
Name: population, dtype: float64

In [16]:
obj = Series([4,7,-5,3])
obj

0    4
1    7
2   -5
3    3
dtype: int64

In [48]:
obj.index = ['Bob', 'Steve', 'Jeff', 'Ryan']
obj

Bob      4
Steve    7
Jeff    -5
Ryan     3
dtype: int64

## 연습문제

학생1은 국어, 영어, 수학 점수가 있고, 학생2는 영어, 수학, 과학 점수가 있다. 점수는 [0,100]의 범위로, 임의의 값이다. 다음을 1) 일반 파이썬 자료구조로 프로그래밍한 것과 2) pandas.Series를 사용해 해결한 것을 비교해 보시오.

1. 각 학생의 공통 과목 점수를 출력하시오.

2. 학생2의 점수 중 학생1은 없는 점수만을 출력하시오.

3. 과목 점수가 같은 것을 찾아, 동점이 있는 과목을 출력하시오.

In [33]:
# 일반 파이썬으로 수행하기
s1 = {u'국어':90, u'수학': 95, u'영어':85}
s2 = {u'영어':75, u'수학': 84, u'과학':50}

# 각 학생의 과목 목록 가져오기
s1_subjects = s1.keys()
s2_subjects = s2.keys()

# 공통 과목 알아내기 
# set 형으로 바꿔 교집합을 구한다.(교재 p.564 세트 참조)
common_sub = set(s1_subjects) & set(s2_subjects)
# 교집합을 구하고 나면 결과는 다시 list로 변환
common_sub = list(common_sub)

# 각 학생의 점수에서, 공통 과목 점수 출력
for s in [s1, s2]:
    print(id(s))
    # 공통 과목 출력
    for subject in common_sub:
        if subject in s:
            print(u'{}: {}'.format(subject, s[subject]))

4365960368
수학: 95
영어: 85
4406405392
수학: 84
영어: 75


In [45]:
# pandas로 수행하기
s1 = Series({u'국어':90, u'수학': 95, u'영어':85})
s2 = Series({u'영어':75, u'수학': 84, u'과학':50})

# 공통 과목만 표시
common_idx = s1.index & s2.index

print(common_idx)
print(s1[common_idx])
print(s2[common_idx])

Index([u'수학', u'영어'], dtype='object')
수학    95
영어    85
dtype: int64
수학    84
영어    75
dtype: int64


점수가 같은 과목만 추출

In [46]:
s1 = Series({u'국어':90, u'수학': 95, u'영어':85})
s2 = Series({u'영어':85, u'과학':50, u'수학': 95,})

# 동점 과목명 구하기
# 불리언 필터 사용 고려 (p.143-145)

In [47]:
s1 == s2

국어    False
수학     True
영어     True
dtype: bool

In [39]:
s1.index

Index([u'국어', u'수학', u'영어'], dtype='object')

In [32]:
s2 == s1

과학    False
수학     True
영어     True
dtype: bool

In [40]:
s2.index

Index([u'과학', u'수학', u'영어'], dtype='object')

## ISSUE?

In [52]:
ss1 = Series([90, 95, 85], index=[u'국어', u'수학', u'영어'])
ss2 = Series([85, 50, 95], index=[u'영어', u'과학', u'수학'])

In [53]:
ss1 == ss2

국어    False
수학    False
영어    False
dtype: bool

In [54]:
ss1[u'수학'] == ss2[u'수학']

True

In [37]:
ss1.index

Index([u'국어', u'수학', u'영어'], dtype='object')

In [38]:
ss2.index

Index([u'영어', u'과학', u'수학'], dtype='object')

# DataFrame

In [22]:
from pandas import Series, DataFrame
DataFrame

pandas.core.frame.DataFrame

In [12]:
data = {u'state': 
        ['Ohio', 'Ohio', 'Ohio', 'Nevada', 'Nevada'],
       'year': [2000, 2001, 2002, 2001, 2002],
       'pop': [1.5, 1.7, 3.6, 2.4, 2.9]}
frame = DataFrame(data)
frame

pop   state  year
0  1.5    Ohio  2000
1  1.7    Ohio  2001
2  3.6    Ohio  2002
3  2.4  Nevada  2001
4  2.9  Nevada  2002

열을 특정한 방식으로 구성하고 싶으면, columns를 설정하면 된다.

In [303]:
frame = DataFrame(data, 
                  columns=['year', 'state', 'pop'])
frame

year   state  pop
0  2000    Ohio  1.5
1  2001    Ohio  1.7
2  2002    Ohio  3.6
3  2001  Nevada  2.4
4  2002  Nevada  2.9

In [13]:
frame2 = DataFrame(data, 
                   columns=['year', 'state', 'pop', 'debt'],
                  index=['하나', '둘', '셋', '넷', '다섯'])
frame2

year   state  pop debt
하나  2000    Ohio  1.5  NaN
둘   2001    Ohio  1.7  NaN
셋   2002    Ohio  3.6  NaN
넷   2001  Nevada  2.4  NaN
다섯  2002  Nevada  2.9  NaN

열(columns)은 추가할 수 있지만, 인덱스는 다른 방식으로 추가해야 한다.

In [14]:
DataFrame(data, columns=['year', 'state', 'pop', 'debt'],
    index=['하나', '둘', '셋', '넷', '다섯', 'SIX'])

ValueError: Shape of passed values is (4, 5), indices imply (4, 6)

In [72]:
frame2.index

Index([u'하나', u'둘', u'셋', u'넷', u'다섯'], dtype='object')

In [63]:
frame2.columns

Index([u'year', u'state', u'pop', u'debt'], dtype='object')

특정 열 선택

In [78]:
frame2['state']

하나      Ohio
둘       Ohio
셋       Ohio
넷     Nevada
다섯    Nevada
Name: state, dtype: object

In [77]:
frame2.state

하나      Ohio
둘       Ohio
셋       Ohio
넷     Nevada
다섯    Nevada
Name: state, dtype: object

In [65]:
type(frame2['state'])

pandas.core.series.Series

행 선택

In [16]:
frame2.ix['셋']

year     2002
state    Ohio
pop       3.6
debt      NaN
Name: 셋, dtype: object

In [78]:
type(frame2.ix['셋'])

pandas.core.series.Series

주의! 유니코드와 일반 문자열은 다르게 취급된다.

In [15]:
frame.ix[u'셋']

KeyError: u'\uc14b'

### 브로드캐스팅
반복문으로 각 행을 반복하며 값을 설정하지 않아도 된다.

In [81]:
frame2.debt = 16.5
frame2

year   state  pop  debt
하나  2000    Ohio  1.5  16.5
둘   2001    Ohio  1.7  16.5
셋   2002    Ohio  3.6  16.5
넷   2001  Nevada  2.4  16.5
다섯  2002  Nevada  2.9  16.5

In [28]:
frame2['debt'] = range(5)
frame2

year   state  pop  debt
하나  2000    Ohio  1.5     0
둘   2001    Ohio  1.7     1
셋   2002    Ohio  3.6     2
넷   2001  Nevada  2.4     3
다섯  2002  Nevada  2.9     4

In [119]:
val = Series([-1.2, -1.5, -1.7, -1.8], 
             index=['둘', '넷', '다섯', '여섯'])
frame2['debt'] = val
frame2

year   state  pop  debt
하나  2000    Ohio  1.5   NaN
둘   2001    Ohio  1.7  -1.2
셋   2002    Ohio  3.6   NaN
넷   2001  Nevada  2.4  -1.5
다섯  2002  Nevada  2.9  -1.7

칼럼 추가는 값 할당으로 가능

In [29]:
frame2.state == 'Ohio'

하나     True
둘      True
셋      True
넷     False
다섯    False
Name: state, dtype: bool

In [120]:
frame2[u'동부'] = frame2.state == 'Ohio'
frame2

year   state  pop  debt     동부
하나  2000    Ohio  1.5   NaN   True
둘   2001    Ohio  1.7  -1.2   True
셋   2002    Ohio  3.6   NaN   True
넷   2001  Nevada  2.4  -1.5  False
다섯  2002  Nevada  2.9  -1.7  False

열 삭제. 간단한 방식. 그러나 이후에 더 나은 방식이 있음.

In [121]:
del frame2[u'동부']
frame2

year   state  pop  debt
하나  2000    Ohio  1.5   NaN
둘   2001    Ohio  1.7  -1.2
셋   2002    Ohio  3.6   NaN
넷   2001  Nevada  2.4  -1.5
다섯  2002  Nevada  2.9  -1.7

그런데, 인덱스는 다음의 방식으로 삭제 불가

In [149]:
del frame2.ix['다섯']

AttributeError: __delitem__

In [151]:
pop = {'Nevada': {2001: 2.4, 2002: 2.9},
      'Ohio':{2000:1.5, 2001:1.7, 2002: 3.6}}
frame3 = DataFrame(pop)
frame3

Nevada  Ohio
2000     NaN   1.5
2001     2.4   1.7
2002     2.9   3.6

In [152]:
frame3.T

2000  2001  2002
Nevada   NaN   2.4   2.9
Ohio     1.5   1.7   3.6

없는 인덱스를 추가하면, NaN이 설정된다.

In [153]:
DataFrame(pop, index=[2001,2002,2003])

Nevada  Ohio
2001     2.4   1.7
2002     2.9   3.6
2003     NaN   NaN

In [157]:
pdata = {'Ohio': frame3['Ohio'][:-1],
        'Nevada': frame3['Nevada'][:2]}
DataFrame(pdata)

Nevada  Ohio
2000     NaN   1.5
2001     2.4   1.7

#### 도전과제

학생은 각각 국어, 수학, 과학의 점수를 갖는다. 학생 세 명의 점수를 DataFrame으로 표현하시오.

1. 모든 학생의 수학 점수를 출력하시오.

2. 각 학생의 과목 평균 입력을 위해 '평균' 열을 추가하시오. 평균값은 아직 계산하지 않아서, 초기값을 0이 되도록 한다.

3. 과목 점수는 국어, 수학, 과학의 순으로 표시하시오.

4. 각 학생의 성별 정보를 '남자' 또는 '여자'로 지정한다. 성별 정보를 활용해 새 열 '여성'을 추가하고, 성별 정보에 따라 True 또는 False의 값을 갖도록 한다. 그리고 문자열로 지정된 원래의 '성별' 열은 삭제한다.

In [18]:
scores = {u'학생1': {u'국어': 70, u'수학': 80, u'과학': 90} , 
          u'학생2': {u'수학': 90, u'국어': 80, u'과학': 100}, 
          u'학생3': {u'과학': 70,u'국어': 90, u'수학': 80}}

frame = DataFrame(scores, index=[u'국어', u'수학', u'과학'])
frame

학생1  학생2  학생3
국어   70   80   90
수학   80   90   80
과학   90  100   70

In [21]:
scores = {u'학생1': {u'국어': 70, u'수학': 80, u'과학': 90, u'성별': u'남'} , 
          u'학생2': {u'수학': 90, u'국어': 80, u'과학': 100, u'성별': u'여'}, 
          u'학생3': {u'과학': 70,u'국어': 90, u'수학': 80, u'성별': u'여'}}

frame = frame.T # 행-열 바꾸기 (p.137 참조)
frame['평균'] = 0
frame.T

/Users/seongjoo/anaconda/lib/python2.7/site-packages/pandas/core/index.py:914: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  return key in self._engine
/Users/seongjoo/anaconda/lib/python2.7/site-packages/pandas/core/index.py:1437: UnicodeWarning: Unicode equal comparison failed to convert both arguments to Unicode - interpreting them as being unequal
  return self._engine.get_loc(_values_from_object(key))


국어  수학   과학 평균     성별
학생1  70  80   90  0   True
학생2  80  90  100  0  False
학생3  90  80   70  0   True
평균    0   0    0  0      0

In [20]:
frame[u'성별'] = [True, False, True]
frame[u'여성'] = frame[u'성별'] == u'여'
del frame[u'성별']
#frame = frame.T
frame

TypeError: invalid type comparison

In [112]:
scores2 = {u'국어': {u'학생1': 70, u'학생2': 80, u'학생3': 90},
          u'수학': {u'학생1': 80, u'학생2':90, u'학생3': 80},
          u'과학': {u'학생1': 90, u'학생2':100, u'학생3': 70}}
frame2 = DataFrame(scores2)
frame2

과학  국어  수학
학생1   90  70  80
학생2  100  80  90
학생3   70  90  80

In [115]:
scores = {u'학생1': [70, 80, 90],
         u'학생2': [80, 90, 100],
         u'학생3': [90, 80]} # 오류! 개수가 맞아야 함.
frame = DataFrame(scores, index=[u'국어', u'수학', u'과학'])
frame

ValueError: could not broadcast input array from shape (2) into shape (3)

## 도전과제

시험 대상 과목 전체는 국어, 수학, 과학, 영어, 프로그래밍이다. 학생들은 세 과목 이상의 점수가 필요하지만 전체 과목을 다 시험볼 필요는 없다. 시험 기준을 충족하는 학생 3명의 정보를 DataFrame으로 구성한다. 이 때, DataFrame에는 모든 과목의 열이 포함되어야 한다.

## 색인 객체

In [159]:
obj = Series(range(3), index=['a', 'b', 'c'])
obj

a    0
b    1
c    2
dtype: int64

In [160]:
index = obj.index
index

Index([u'a', u'b', u'c'], dtype='object')

In [91]:
index[0]

'a'

In [90]:
index[1:]

Index([u'b', u'c'], dtype='object')

In [93]:
index[1] = 'a' # 오류!

TypeError: Indexes does not support mutable operations

In [161]:
obj2 = Series([1.5, -2.5, 0], index=index)
obj2

a    1.5
b   -2.5
c    0.0
dtype: float64

In [162]:
obj.index is obj2.index

True

## 재색인

In [30]:
obj = Series([4.5, 7.2, -5.3, 3.6], 
             index=['d', 'b', 'a', 'c'])
obj

d    4.5
b    7.2
a   -5.3
c    3.6
dtype: float64

객체가 생성된 이후 인덱스 조작

In [31]:
obj2 = obj.reindex(['a','b','c','d','e'])
obj2 # 인덱스 e 추가

a   -5.3
b    7.2
c    3.6
d    4.5
e    NaN
dtype: float64

.reindex() --> 새로운 객체가 반환된다.

In [163]:
obj is obj2

False

In [98]:
obj.reindex(['a','b','c','d','e'], fill_value=0)

a   -5.3
b    7.2
c    3.6
d    4.5
e    0.0
dtype: float64

In [22]:
obj3 = Series(['blue', 'purple', 'yellow'], 
              index=[0,2,4])
obj3

0      blue
2    purple
4    yellow
dtype: object

중간을 채워주는 보간(interpolation)

In [24]:
obj3.reindex(range(6), method='ffill')

0      blue
1      blue
2    purple
3    purple
4    yellow
5    yellow
dtype: object

In [167]:
obj3.reindex(range(6), 
             method='bfill')

0      blue
1    purple
2    purple
3    yellow
4    yellow
5       NaN
dtype: object

DataFrame 재색인

In [36]:
import numpy as np

data = np.arange(9).reshape((3,3))
frame = DataFrame(data, index=['a', 'c', 'd'], 
                  columns=['Ohio', 'Texas', 'California'])
frame

Ohio  Texas  California
a     0      1           2
c     3      4           5
d     6      7           8

재색인을 통해 인덱스 추가

In [172]:
frame2 = frame.reindex(['a','b','c','d'])
frame2

Ohio  Texas  California
a     0      1           2
b   NaN    NaN         NaN
c     3      4           5
d     6      7           8

인덱스 뿐만 아니라 열을 재색인할 수도 있다.

In [174]:
states = ['Texas', 'Utah', 'California']
frame.reindex(columns=states)

Texas  Utah  California
a      1   NaN           2
c      4   NaN           5
d      7   NaN           8

In [175]:
frame.reindex(index=['a', 'b', 'c', 'd'], method='ffill', columns=states)

Texas  Utah  California
a      1   NaN           2
b      1   NaN           2
c      4   NaN           5
d      7   NaN           8

In [176]:
frame.ix[['a', 'b', 'c', 'd'], states]

Texas  Utah  California
a      1   NaN           2
b    NaN   NaN         NaN
c      4   NaN           5
d      7   NaN           8

#### 도전과제

성주는 SJ전자와 매크로소프트 사의 주식을 가지고 있다. 

1. 각 주식의 10일 간의 가격을 DataFrame으로 표현하시오. 가격 정보는 임의로 설정하시오.

2. 메론 사의 주식 정보를 DataFrame에 추가하시오.

3. 매크로소프트, 메론, SJ전자 순으로 회사명이 행 방향으로 표시되도록 하시오.

4. 코드베이직 사는 5일 전에 새로 상장되었다. 그래서 지난 5일 간의 가격만 있고, 이전 가격은 없다. 코드베이직 사의 가격을 DataFrame에 추가하시오. (참조: p.170, In[50])

5. 모든 주식에 대해 가격 정보가 있는 구간만을 선택하시오. (참조: p.181-182)

6. 코드베이직과 SJ 전자의 6일부터 8일까지의 가격만을 선택하시오. (참조: p.182-183)

In [55]:
# 문제 1
stock_prices = np.random.randint(100,200, 
                                 size=(10,2))
frame = DataFrame(stock_prices, 
                  columns=['SJ전자', '매크로소프트'])
frame

SJ전자  매크로소프트
0   163     120
1   119     128
2   132     129
3   101     183
4   177     171
5   145     106
6   104     127
7   119     167
8   137     181
9   118     184

In [226]:
# 문제 2
frame[u'메론'] = np.random.randint(100,200, 
                                 size=10)
frame

SJ전자  매크로소프트   메론
0   116     125  150
1   107     142  120
2   156     150  186
3   147     110  100
4   154     145  129
5   158     198  165
6   177     148  137
7   102     128  180
8   188     138  114
9   132     110  108

In [200]:
# 문제 c
frame.reindex(columns=[u'매크로소프트', u'메론', u'SJ전자']).T

0    1    2    3    4    5    6    7    8    9
매크로소프트  104  193  194  146  168  197  135  144  197  192
메론      149  161  100  140  177  186  148  139  192  105
SJ전자    137  192  139  138  133  157  178  140  134  101

문제 4

In [57]:
frame[u'코드베이직'] = Series(np.random.randint(100,200, size=5)
                        ,index=[5,6,7,8,9])
frame

SJ전자  매크로소프트  코드베이직
0   163     120    NaN
1   119     128    NaN
2   132     129    NaN
3   101     183    NaN
4   177     171    NaN
5   145     106    114
6   104     127    140
7   119     167    124
8   137     181    119
9   118     184    130

In [243]:
frame.notnull()

SJ전자 매크로소프트    메론  코드베이직
0  True   True  True  False
1  True   True  True  False
2  True   True  True  False
3  True   True  True  False
4  True   True  True  False
5  True   True  True   True
6  True   True  True   True
7  True   True  True   True
8  True   True  True   True
9  True   True  True   True

In [244]:
frame.notnull().all()

SJ전자       True
매크로소프트     True
메론         True
코드베이직     False
dtype: bool

In [242]:
frame.ix[5:8, [u'코드베이직', u'SJ전자']]

코드베이직  SJ전자
5    169   158
6    100   177
7    158   102
8    192   188

행열 삭제

In [58]:
obj=Series(np.arange(5.), index=['a','b','c','d','e'])
obj

a    0
b    1
c    2
d    3
e    4
dtype: float64

In [59]:
obj.drop('c')

a    0
b    1
d    3
e    4
dtype: float64

In [115]:
obj.drop(['d', 'c'])

a    0
b    1
e    4
dtype: float64

In [60]:
data = np.arange(16).reshape((4,4)) # 4x4 행렬
data

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

In [61]:
frame = DataFrame(data, 
                  index=['Ohio', 'Colorado', 'Utah', 'New York'],
                 columns=['one','two', 'trois', 'quatre'])
frame

one  two  trois  quatre
Ohio        0    1      2       3
Colorado    4    5      6       7
Utah        8    9     10      11
New York   12   13     14      15

In [62]:
frame.drop(['Colorado', 'Ohio'])

one  two  trois  quatre
Utah        8    9     10      11
New York   12   13     14      15

열 삭제

In [123]:
frame.drop(['one','two'], axis=1)

trois  quatre
Ohio          2       3
Colorado      6       7
Utah         10      11
New York     14      15

원판, ... 아니 원본은 불변

In [63]:
frame

one  two  trois  quatre
Ohio        0    1      2       3
Colorado    4    5      6       7
Utah        8    9     10      11
New York   12   13     14      15

### 색인, 선택, 필터

obj[<선택조건>]

In [204]:
obj = Series(np.arange(4.0), index=list('abcd'))
obj

a    0
b    1
c    2
d    3
dtype: float64

In [203]:
obj['b']

1.0

In [205]:
obj[1]

1.0

In [206]:
obj[2:4]

c    2
d    3
dtype: float64

In [207]:
obj[['b','a','d']]

b    1
a    0
d    3
dtype: float64

In [208]:
obj[[1,3]]

b    1
d    3
dtype: float64

In [209]:
obj[obj<2]

a    0
b    1
dtype: float64

문자 라벨 슬라이싱

In [64]:
obj = Series(range(5), index=['d','e','b','a','c'])
obj

d    0
e    1
b    2
a    3
c    4
dtype: int64

In [65]:
obj['b':'a']

b    2
a    3
dtype: int64

In [67]:
obj['b' : 'c']

b    2
a    3
c    4
dtype: int64

In [68]:
obj['b':'d']

Series([], dtype: int64)

In [16]:
obj[2:4]

d    0
e    1
b    2
a    3
c    4
dtype: int64
b    2
a    3
dtype: int64
b    2
a    3
dtype: int64


In [212]:
np.arange(16).reshape((4,4))

array([[ 0,  1,  2,  3],
       [ 4,  5,  6,  7],
       [ 8,  9, 10, 11],
       [12, 13, 14, 15]])

In [70]:
data = DataFrame(np.arange(16).reshape((4,4)),
                index=['Ohio', 'Colorado', 'Utah', 'New York'],
                columns=['one', 'two', 'three', 'four'])
data

one  two  three  four
Ohio        0    1      2     3
Colorado    4    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

In [214]:
data['two']

Ohio         1
Colorado     5
Utah         9
New York    13
Name: two, dtype: int64

In [215]:
data[['three', 'one']]

three  one
Ohio          2    0
Colorado      6    4
Utah         10    8
New York     14   12

질문: 열 방향으로 선택되어야 하는 거 아냐?

In [71]:
data[:2]

one  two  three  four
Ohio        0    1      2     3
Colorado    4    5      6     7

In [217]:
data[data['three'] > 5]

one  two  three  four
Colorado    4    5      6     7
Utah        8    9     10    11
New York   12   13     14    15

행열 방향 색인과 슬라이싱

In [218]:
data.ix['Colorado', ['two', 'three']]

two      5
three    6
Name: Colorado, dtype: int64

In [219]:
data.ix[['Colorado', 'Utah'], [3,0,1]]

four  one  two
Colorado     7    4    5
Utah        11    8    9

In [220]:
data.ix[2]

one       8
two       9
three    10
four     11
Name: Utah, dtype: int64

In [221]:
data.ix[:'Utah', 'two']

Ohio        1
Colorado    5
Utah        9
Name: two, dtype: int64

In [73]:
data[:2]

one  two  three  four
Ohio        0    1      2     3
Colorado    4    5      6     7

In [74]:
data.ix[:2, :]

one  two  three  four
Ohio        0    1      2     3
Colorado    4    5      6     7

In [72]:
data.ix[:, :2]

one  two
Ohio        0    1
Colorado    4    5
Utah        8    9
New York   12   13

In [222]:
data.ix[data.three > 5, :3]

one  two  three
Colorado    4    5      6
Utah        8    9     10
New York   12   13     14

#### 도전과제

학생은 각각 국어, 수학, 과학의 점수를 갖는다. 학생 세 명의 점수를 DataFrame으로 표현하시오.

1. 학생1과 학생3의 수학, 과학 점수만을 선택하시오.

2. 수학 점수가 80점 이상인 학생들만 표시하시오.

In [89]:
scores = DataFrame(np.random.randint(0,101, size=(3,3)),
                  index=[u'국어',u'수학',u'과학'],
                  columns=[u'학생1', u'학생2', u'학생3'])
scores

학생1  학생2  학생3
국어   29   69   71
수학   57   88   47
과학   83    2   83

In [97]:
selection = scores.ix[u'수학']>=80
selection

학생1    False
학생2     True
학생3    False
Name: 수학, dtype: bool

In [98]:
scores[selection]

IndexingError: Unalignable boolean Series key provided

In [91]:
selection = scores.ix[u'수학']>=80
scores[selection.index]

학생1  학생2  학생3
국어   29   69   71
수학   57   88   47
과학   83    2   83

In [94]:
selection.values

array([False,  True, False], dtype=bool)

In [99]:
scores[selection.values]

학생1  학생2  학생3
수학   57   88   47

수학 점수를 기준으로 누가 기준을 넘는지 확인

In [96]:
scores.ix[:, [False, True, False]]

학생2
국어   69
수학   88
과학    2

In [101]:
scores.ix[:, scores.ix[u'수학']>=80]

학생2
국어   69
수학   88
과학    2

## 산술연산

In [102]:
s1 = Series([7.3, -2.5, 3.4, 1.5], 
            index=['a','c','d','e'])
s2 = Series([-2.1, 3.6, -1.5, 4, 3.1], 
            index=['a','c','e','f','g'])

In [103]:
s1

a    7.3
c   -2.5
d    3.4
e    1.5
dtype: float64

In [104]:
s2

a   -2.1
c    3.6
e   -1.5
f    4.0
g    3.1
dtype: float64

In [105]:
s1+s2

a    5.2
c    1.1
d    NaN
e    0.0
f    NaN
g    NaN
dtype: float64

In [106]:
df1 = DataFrame(np.arange(12.).reshape((3,4)),
              columns=list('abcd'))
df2 = DataFrame(np.arange(20.).reshape((4,5)), 
               columns=list('abcde'))

In [107]:
df1

a  b   c   d
0  0  1   2   3
1  4  5   6   7
2  8  9  10  11

In [108]:
df2

a   b   c   d   e
0   0   1   2   3   4
1   5   6   7   8   9
2  10  11  12  13  14
3  15  16  17  18  19

In [109]:
df1+df2

a   b   c   d   e
0   0   2   4   6 NaN
1   9  11  13  15 NaN
2  18  20  22  24 NaN
3 NaN NaN NaN NaN NaN

NaN 대신 기본값 설정하기

In [28]:
df1.add(df2, fill_value=0)

a   b   c   d   e
0   0   2   4   6   4
1   9  11  13  15   9
2  18  20  22  24  14
3  15  16  17  18  19

DataFrame과 Series 연산

In [111]:
frame = DataFrame(np.random.randn(4,3), 
               columns=list('bde'),
              index=['Utah', 'Ohio', 'Texas', 'Oregon'])

series = frame.ix[0]

print(frame)
print(frame - series)

               b         d         e
Utah    0.231691 -0.329679  1.264903
Ohio    1.664624 -0.137734  0.996886
Texas  -0.026082  0.795804  0.700870
Oregon  1.669533  0.058905  1.351678
b    0.231691
d   -0.329679
e    1.264903
Name: Utah, dtype: float64
               b         d         e
Utah    0.000000  0.000000  0.000000
Ohio    1.432933  0.191945 -0.268017
Texas  -0.257772  1.125483 -0.564033
Oregon  1.437842  0.388584  0.086776


In [113]:
series2 = Series(range(1,4), index=['b','e','f'])
series2

b    1
e    2
f    3
dtype: int64

In [114]:
frame+series2

b   d         e   f
Utah    1.231691 NaN  3.264903 NaN
Ohio    2.664624 NaN  2.996886 NaN
Texas   0.973918 NaN  2.700870 NaN
Oregon  2.669533 NaN  3.351678 NaN

In [115]:
series3 = frame['d']
series3

Utah     -0.329679
Ohio     -0.137734
Texas     0.795804
Oregon    0.058905
Name: d, dtype: float64

series의 인덱스는 frame의 columns로 간주하고 연산 수행 (기본)

In [39]:
frame - series3

Ohio  Oregon  Texas  Utah   b   d   e
Utah     NaN     NaN    NaN   NaN NaN NaN NaN
Ohio     NaN     NaN    NaN   NaN NaN NaN NaN
Texas    NaN     NaN    NaN   NaN NaN NaN NaN
Oregon   NaN     NaN    NaN   NaN NaN NaN NaN

series의 인덱스를 frame의 인덱스로 간주하고 연산 수행

In [119]:
series3 = frame['d']
series3

Utah     -0.329679
Ohio     -0.137734
Texas     0.795804
Oregon    0.058905
Name: d, dtype: float64

In [121]:
frame.sub(series3, axis=1)

Ohio  Oregon  Texas  Utah   b   d   e
Utah     NaN     NaN    NaN   NaN NaN NaN NaN
Ohio     NaN     NaN    NaN   NaN NaN NaN NaN
Texas    NaN     NaN    NaN   NaN NaN NaN NaN
Oregon   NaN     NaN    NaN   NaN NaN NaN NaN

## 도전과제

10명의 학생이 5개의 과목 점수를 갖는다. 과목 점수는 [0,100] 범위이다. 과목명은 숫자 코드로 되어 있으며, 숫자코드는 [101, 105] 구간의 값이다. 

1. 과목코드 102의 성적이 잘못 매겨졌다. 전체 10명의 학생은 재채점 결과에 따라 각각 [-10,10] 범위의 임의 점수가 현재 점수에 반영되도록 한다.

2. 과목 점수 재조정 시, 점수가 음수가 되거나, 100을 초과하지 않도록 하시오.

In [128]:
scores = np.random.randint(0,101, size=(10,5)) # 10x5
frame_scores = DataFrame(scores, 
                         index=range(1,11),
                         columns=range(101, 106))
frame_scores

101  102  103  104  105
1    14   25   87   85   39
2    61   89   32   38   92
3    59   73   42   28   57
4    41   40   90   27   33
5    31   35   10   68   19
6    63    3   63   59    3
7    35   46   11   25    3
8    71   12   31   60    2
9    18   22   11   89   17
10   26   84   98   23   71

In [129]:
# a
deltas = np.random.randint(-10, 10, size=10)
score_delta = Series(deltas, index=frame_scores.index)
score_delta

1    -7
2     7
3    -3
4    -7
5    -1
6     6
7    -1
8     8
9    -7
10   -2
dtype: int64

방법 1. 특정 열을 선택해서 연산을 수행한 다음, 해당 열을 갱신

In [130]:
frame_scores[102] = frame_scores[102] + score_delta
frame_scores

101  102  103  104  105
1    14   18   87   85   39
2    61   96   32   38   92
3    59   70   42   28   57
4    41   33   90   27   33
5    31   34   10   68   19
6    63    9   63   59    3
7    35   45   11   25    3
8    71   20   31   60    2
9    18   15   11   89   17
10   26   82   98   23   71

방법 2. DataFrame 간의 연산으로 변환해서 수행

In [138]:
new_frame_scores = frame_scores.sub(
    DataFrame(score_delta, columns=[102]), fill_value=0)

문제 2 해결

In [142]:
DataFrame(np.where(new_frame_scores < 0 , 0, 
                  np.where(new_frame_scores>100, 100, new_frame_scores)),
         index = new_frame_scores.index,
         columns = new_frame_scores.columns)

101  102  103  104  105
1    14   25   87   85   39
2    61   89   32   38   92
3    59   73   42   28   57
4    41   40   90   27   33
5    31   35   10   68   19
6    63    3   63   59    3
7    35   46   11   25    3
8    71   12   31   60    2
9    18   22   11   89   17
10   26   84   98   23   71

##  함수 적용과 매핑

In [307]:
import numpy as np
frame = DataFrame(np.random.randn(4,3), 
                 columns=list('bde'),
                 index=['Utah', 'Ohio', 'Texas', 'Oregon'])

In [144]:
np.abs(frame)

b         d         e
Utah    0.443940  0.226324  0.186033
Ohio    1.548336  1.358102  1.053157
Texas   0.342912  1.610758  0.259303
Oregon  0.287597  0.506503  0.311234

In [145]:
frame.apply(np.abs)

b         d         e
Utah    0.443940  0.226324  0.186033
Ohio    1.548336  1.358102  1.053157
Texas   0.342912  1.610758  0.259303
Oregon  0.287597  0.506503  0.311234

In [308]:
def maxmindiff(x):
    return x.max() - x.min()

In [309]:
frame.apply(maxmindiff)

b    2.896893
d    1.895672
e    4.626412
dtype: float64

In [310]:
frame.apply(lambda x : x.max()- x.min())

b    2.896893
d    1.895672
e    4.626412
dtype: float64

In [149]:
frame.apply(maxmindiff, axis=1)

Utah      0.670263
Ohio      2.601493
Texas     1.351455
Oregon    0.817737
dtype: float64

In [311]:
def get_max_min(x):
    return Series([x.min(), x.max()], index=['min', 'max'])

In [312]:
frame.apply(get_max_min)

b         d         e
min -2.430589 -1.271455 -2.480493
max  0.466304  0.624217  2.145918

In [152]:
frame.apply(f, axis=1)

min       max
Utah   -0.443940  0.226324
Ohio   -1.053157  1.548336
Texas  -1.610758 -0.259303
Oregon -0.506503  0.311234

In [314]:
frame.applymap(lambda x: '{:.2f}'.format(x))

b      d      e
Utah    -2.43   0.48  -0.51
Ohio    -0.36  -1.27   2.15
Texas    0.25  -0.49  -2.48
Oregon   0.47   0.62   0.02

## 도전과제

주식 종목 5개 코드베이직, SJ전자, 매크로소프트, 메론, 성주모직의 90일간의 가격 정보에 대해 다음을 수행한다. 각 주식의 가격은 천원 단위로 기록되어 있다.

1. 각 종목의 평균 가격 산출

2. 각 종목의 최고가, 최저가 산출

3. 모든 종목의 당일 평균 가격 산출

4. 천원 단위로 기록된 수치값을 원 단위로 변경해 단위 구분자(,)를 포함해 표시한다. 예: 15 --> 15,000

5. 종목 가격 정보를 문자의 오름차순으로 정렬해 표시한다.

6. 각 주식의 최고값과 최저값의 차이를 구하고, 차이가 큰 종목 순으로 내림차순으로 정렬해 표시한다.

7. 90일간 가격의 편차가 큰 순으로 순위를 매긴다.

8. 각 종목 간 가격 변화량 상관도 (correlation)을 구한다. 

9. 코드베이직과 가장 상관도가 높은 종목을 순서대로 두 개 표시한다.

In [144]:
print('{:,}'.format(1234567890))

1,234,567,890


In [143]:
def gen_sim_price(num_stocks=10, num_days=90):
    """주식 종목 가격 시뮬레이션. 
    지정된 종목 개수에 대한 지정된 일수만큼의 모의 가격 생성"""
    prices = np.random.randint(5,1000, size=num_stocks)
    changes = 1+np.random.uniform(low=-0.3, high=0.3, 
                                  size=(num_days, num_stocks))
    prices =  prices * changes.cumprod(0)

    return prices

prices = gen_sim_price(5) # 5개 종목 90일 가격 생성
stock_names = [u'코드베이직', u'SJ전자', u'매크로소프트', u'메론', u'성주모직']
price_frame = DataFrame(prices,
                       columns=stock_names)
price_frame

코드베이직         SJ전자     매크로소프트           메론        성주모직
0   40.772264  1137.567283  38.567574   916.032626  330.057412
1   36.265453  1160.236864  37.334749  1042.277046  264.196470
2   31.290528   973.582639  27.145672   805.160861  233.021844
3   40.386841  1069.696779  19.113824   853.205257  293.966419
4   45.566729   762.815341  15.105072   935.009369  253.566603
5   58.581456   624.773357  13.379328  1203.061379  306.486700
6   65.357062   498.925705   9.402117  1093.106242  341.917008
7   79.198429   583.471045  10.155532  1106.151972  258.235495
8   72.470239   470.581131   9.782670  1423.551258  214.839002
9   53.305959   562.680954  12.539343  1066.057610  213.852970
10  51.241655   490.991859  10.123904   986.325130  214.501398
11  40.743620   486.855549  10.755345   913.587077  167.167196
12  44.166414   518.869882  10.223529   872.023135  163.798869
13  48.416910   417.232910  10.161343   815.228473  116.280727
14  56.083500   333.567627   9.963379   911.863977  137.290767
15  59.243399   272.976505   9.669228   818.551682  103.208626
16  64.041175   256.087943   6.955028  1014.539806  116.635226
17  67.513519   205.952544   8.245207  1022.389078  142.679212
18  70.646894   256.855793   6.087057  1254.892050  150.847938
19  66.982525   220.274710   4.602781   960.081002  191.900610
20  64.148421   198.223814   3.285512  1153.744864  230.242928
21  75.745803   230.044549   4.172996  1089.079427  281.528132
22  92.384601   262.899934   5.419655  1349.051978  362.341492
23  74.747906   215.412352   6.590344  1316.066179  302.259655
24  74.620222   164.127665   7.238685  1184.304512  260.760708
25  62.337775   210.014574   9.357892  1310.058944  332.003547
26  59.149157   262.447479   8.480521  1416.162435  271.104795
27  48.828476   263.568055   6.135224  1104.400037  213.754197
28  60.851628   209.417847   5.552597  1016.946630  183.819062
29  77.212343   225.878091   5.506449  1092.544840  175.419692
..        ...          ...        ...          ...         ...
60  37.018527   269.288241   3.114258  1453.002954   97.200768
61  33.372594   279.835412   2.956046  1754.812498  102.085888
62  31.150173   304.413452   2.675593  1729.993433  103.869287
63  37.668114   321.071220   3.406275  1457.996289  127.608214
64  39.650258   292.584375   2.407836  1672.721051   91.604993
65  40.808004   232.069492   2.754303  1693.838408  101.435146
66  30.314623   284.558644   3.172315  1667.335756  101.760911
67  26.162360   288.739934   2.846685  1192.424871  123.805803
68  26.470834   294.283488   2.123599  1394.315682  127.226238
69  29.077149   232.279524   2.178158  1363.764870  115.423679
70  27.678049   290.881618   1.787127  1333.683209  142.626817
71  30.472794   357.388323   1.721579  1576.655935  123.182636
72  25.865093   411.652747   1.351082  1931.344338  121.623934
73  19.453197   425.275663   1.013039  2392.984969  146.172729
74  17.283220   373.639051   0.827453  2325.566312  102.385559
75  14.752751   282.394202   0.814962  1653.907148  118.942283
76  19.168514   264.992539   0.613250  1362.740279  103.477070
77  13.498341   342.677845   0.666915  1011.116702   86.708973
78  11.574696   280.560530   0.520410   991.188738  101.863360
79  12.015599   274.193181   0.622222  1208.204705  122.616887
80  13.490588   297.343334   0.722365  1232.456951  141.699580
81  11.459370   226.498284   0.738072  1274.778008  136.036251
82  10.005651   234.393873   0.918351  1085.308089  124.177271
83   8.708612   202.075735   0.763790   972.439775  109.256701
84   9.317444   187.142190   0.713027  1000.710236  116.907845
85  11.361603   229.226666   0.788922  1106.249334  120.860039
86  11.652398   185.757031   0.815907  1016.049022   93.894793
87  13.074571   230.037866   0.993182  1284.302871  104.407476
88  11.284322   181.791322   1.193138  1440.051953   87.840473
89  11.442998   156.534287   1.439301  1699.897886  103.326621

[90 rows x 5 columns]

In [361]:
# 각 종목의 평균
price_frame.apply(np.mean)

코드베이직     100.799240
SJ전자       87.655959
매크로소프트    115.634427
메론        240.975945
성주모직      713.977382
dtype: float64

In [362]:
# 각 종목의 최소, 최대값
price_frame.apply(lambda x: Series([x.min(), x.max()], 
                                   index=[u'최소', u'최대']))

코드베이직        SJ전자      매크로소프트           메론         성주모직
최소    9.735483    9.052450   23.097862    39.708412   226.341596
최대  541.590510  483.236142  323.579386  1092.455677  1404.141261

In [363]:
# 일자별 종목 가격 평균
price_frame.apply(np.mean, axis=1)

0     531.452063
1     524.366050
2     534.760620
3     507.869819
4     452.882953
5     499.808172
6     492.450456
7     483.012034
8     422.288022
9     423.084028
10    407.851319
11    380.519480
12    349.640397
13    370.124287
14    387.101247
15    329.070431
16    334.227295
17    382.705393
18    324.330995
19    330.144582
20    297.725144
21    266.026084
22    256.799235
23    257.643812
24    227.322978
25    219.816188
26    230.598961
27    264.351872
28    314.850983
29    290.530870
         ...    
60    114.361000
61    102.091048
62    109.461595
63    113.276930
64    127.172267
65    117.137542
66    113.673275
67    115.520197
68    111.344968
69     97.073207
70     99.037702
71    110.806108
72    131.858295
73    139.070688
74    151.687857
75    170.693087
76    129.141332
77    127.564773
78    139.003045
79    157.772172
80    166.622858
81    205.509489
82    193.405715
83    217.067095
84    241.351202
85    219.898025
86    243.350004
87    284.3655

In [364]:
# 가격 표시를 원으로 변경
price_frame.applymap(lambda x: '{:,}'.format(x*1000))

코드베이직            SJ전자          매크로소프트               메론  \
0    450,603.18408   461,501.14213  72,455.6332916   639,593.107162   
1   501,038.677125  483,236.142292  88,281.3777392    640,925.93916   
2   533,969.958633  393,733.728378  108,115.293246   790,353.234855   
3   472,558.069889  376,394.126782  88,259.3640367   812,189.282906   
4   541,590.510078  318,198.190404  104,568.602954   730,878.014515   
5   482,745.448166  295,398.028904   84,806.129898   915,168.974091   
6   458,519.424988  245,762.577151  110,146.335324   924,807.059386   
7   352,606.152954  256,781.980286  102,340.098169  1,092,455.67662   
8   394,105.191225  239,477.820933  91,990.6363843   777,606.258061   
9   345,109.369301  299,711.909338  113,170.122572   671,081.905563   
10  264,794.251256  242,453.361242  145,343.990303   591,994.691862   
11  246,314.960821  240,300.393605  125,522.995084   474,636.275726   
12  236,982.412152  194,368.991142  161,555.164073   377,684.702629   
13  197,129.157235  246,919.023885  147,052.684678   266,493.529287   
14  154,916.820016  317,007.737657  150,247.388451   278,457.208668   
15  162,511.822817  360,550.223079  113,398.122196   237,021.033317   
16   168,317.98921  256,793.870803  92,421.8933072   209,241.730483   
17  126,469.488354    229,013.3501  88,247.7005464   244,551.677026   
18  118,887.914321  161,509.426543  105,681.525368   185,839.545509   
19  92,080.4387395  151,323.789479  81,812.4543834   216,989.268182   
20  101,249.842393  152,092.411378  65,645.0435923   271,023.948794   
21  127,280.978387  171,372.882124  47,669.6911638   287,745.240009   
22  91,287.5435286  149,709.202709   54,793.330116   205,743.754575   
23  81,157.9709028  113,683.943497  40,948.2064684   252,652.472586   
24  87,790.2238975  97,935.8094383  34,198.3628699   244,655.705511   
25  98,789.2013472  69,279.5828078  41,465.6753557   175,343.445741   
26  117,857.092075  55,523.0194075  49,200.9830013   128,601.410265   
27  83,698.6105844  43,075.6922117   43,117.626636    151,920.99313   
28   93,077.986193  37,458.9393501  42,849.1057797   182,588.872362   
29  94,629.0986262  32,643.0352769  36,407.4674804   196,302.078439   
..             ...             ...             ...              ...   
60   19,564.611991  9,125.49789284  94,511.8686973   58,434.3134721   
61  18,251.3793552  10,010.7400757  121,176.351692   65,811.7317883   
62  18,461.5714135  12,025.6073817  125,516.495832   65,154.8989905   
63   21,917.186914  15,309.2441771  141,574.668699   73,921.0840905   
64  25,655.1398257  11,919.3863679   165,159.68574   55,233.4363515   
65  30,305.7485991  13,710.8509845  117,325.462691   49,324.8653796   
66  30,197.1131389  14,447.8828603  141,180.240171   47,874.4664465   
67  37,342.3802097   11,213.386498  113,898.289389   39,708.4121962   
68   35,682.302834  12,358.4134257   145,632.68853   44,605.3541043   
69  28,379.5892607  13,662.6615276  154,509.401215   44,547.2142894   
70  21,193.8093736  10,877.7943168  192,457.946563   44,317.3626597   
71  19,188.7060483  10,291.7953303  238,285.398066   54,403.6467643   
72  22,140.2431915  12,074.7012845   273,127.60723   70,402.2769992   
73  22,901.2796046  13,569.1165435  222,860.656948   76,060.8452093   
74  26,939.8517738  16,982.8483272  200,457.988292   92,384.0637526   
75  23,581.2903998  21,276.0739207  233,023.546696   86,190.0400307   
76  26,719.8349255  17,937.0487197  170,288.203246   73,044.4636526   
77  20,234.7662874  19,732.3352464  173,425.022485   52,302.6848348   
78  14,631.8024761  24,288.0517346  211,627.767053   59,252.1756786   
79  15,663.5777962  30,617.3578111  194,948.359718   72,414.3817481   
80  13,803.5820226  35,593.4250035  249,149.156252   52,969.3622081   
81  12,467.4400277  32,775.2732776  323,579.385518   64,377.1925483   
82  9,735.48293173  32,311.4707002  282,930.750862   83,662.6466476   
83    12,181.89135  33,157.5566768  306,568.407085   68,317.7758765   
84  13,859.9833656  27,573.6455113  2

In [365]:
# 종목 정보를 문자 순으로 정렬
price_frame.sort_index(axis=1)

SJ전자      매크로소프트           메론         성주모직       코드베이직
0   461.501142   72.455633   639.593107  1033.107247  450.603184
1   483.236142   88.281378   640.925939   908.348115  501.038677
2   393.733728  108.115293   790.353235   847.630886  533.969959
3   376.394127   88.259364   812.189283   789.948253  472.558070
4   318.198190  104.568603   730.878015   569.179448  541.590510
5   295.398029   84.806130   915.168974   720.922277  482.745448
6   245.762577  110.146335   924.807059   723.016881  458.519425
7   256.781980  102.340098  1092.455677   610.876264  352.606153
8   239.477821   91.990636   777.606258   608.260201  394.105191
9   299.711909  113.170123   671.081906   686.346834  345.109369
10  242.453361  145.343990   591.994692   794.670300  264.794251
11  240.300394  125.522995   474.636276   815.822774  246.314961
12  194.368991  161.555164   377.684703   777.610714  236.982412
13  246.919024  147.052685   266.493529   993.027038  197.129157
14  317.007738  150.247388   278.457209  1034.877081  154.916820
15  360.550223  113.398122   237.021033   771.870954  162.511823
16  256.793871   92.421893   209.241730   944.360993  168.317989
17  229.013350   88.247701   244.551677  1225.244749  126.469488
18  161.509427  105.681525   185.839546  1049.736563  118.887914
19  151.323789   81.812454   216.989268  1108.516960   92.080439
20  152.092411   65.645044   271.023949   898.614473  101.249842
21  171.372882   47.669691   287.745240   696.061629  127.280978
22  149.709203   54.793330   205.743755   782.462345   91.287544
23  113.683943   40.948206   252.652473   799.776464   81.157971
24   97.935809   34.198363   244.655706   672.034790   87.790224
25   69.279583   41.465675   175.343446   714.203036   98.789201
26   55.523019   49.200983   128.601410   801.812302  117.857092
27   43.075692   43.117627   151.920993   999.946438   83.698611
28   37.458939   42.849106   182.588872  1218.280010   93.077986
29   32.643035   36.407467   196.302078  1092.672670   94.629099
..         ...         ...          ...          ...         ...
60    9.125498   94.511869    58.434313   390.168710   19.564612
61   10.010740  121.176352    65.811732   295.205037   18.251379
62   12.025607  125.516496    65.154899   326.149401   18.461571
63   15.309244  141.574669    73.921084   313.662466   21.917187
64   11.919386  165.159686    55.233436   377.893687   25.655140
65   13.710851  117.325463    49.324865   375.020785   30.305749
66   14.447883  141.180240    47.874466   334.666673   30.197113
67   11.213386  113.898289    39.708412   375.438517   37.342380
68   12.358413  145.632689    44.605354   318.446083   35.682303
69   13.662662  154.509401    44.547214   244.267171   28.379589
70   10.877794  192.457947    44.317363   226.341596   21.193809
71   10.291795  238.285398    54.403647   231.860994   19.188706
72   12.074701  273.127607    70.402277   281.546646   22.140243
73   13.569117  222.860657    76.060845   359.961541   22.901280
74   16.982848  200.457988    92.384064   421.674535   26.939852
75   21.276074  233.023547    86.190040   489.394484   23.581290
76   17.937049  170.288203    73.044464   357.717108   26.719835
77   19.732335  173.425022    52.302685   372.129054   20.234766
78   24.288052  211.627767    59.252176   385.215426   14.631802
79   30.617358  194.948360    72.414382   475.217182   15.663578
80   35.593425  249.149156    52.969362   481.598767   13.803582
81   32.775273  323.579386    64.377193   594.348154   12.467440
82   32.311471  282.930751    83.662647   558.388224    9.735483
83   33.157557  306.568407    68.317776   665.109845   12.181891
84   27.573646  258.946198    74.387760   831.988422   13.859983
85   28.911356  255.893869    78.102480   726.733793    9.848626
86   30.814713  261.431610    98.752256   815.579359   10.172080
87   32.514819  286.014401    74.316302  1015.999030   12.983335
88   31.947173  227.222638    90.965610  1170.427916    9.985469
89   23.650107  292.388044    78.497692  1404

In [367]:
"""각 주식의 최고값과 최저값의 차이를 구하고, 
차이가 큰 종목 순으로 내림차순으로 정렬해 표시한다."""
price_frame.apply(lambda x: x.max()-x.min()).order()

매크로소프트     300.481523
SJ전자       474.183692
코드베이직      531.855027
메론        1052.747264
성주모직      1177.799666
dtype: float64

In [371]:
"""90일간 가격의 편차가 큰 순으로 순위를 매긴다."""
price_frame.apply(np.std).rank().order()

매크로소프트    1
SJ전자      2
코드베이직     3
메론        4
성주모직      5
dtype: float64

In [408]:
# 종목별 가격 변화량 상관관계
price_frame.pct_change().corr()

코드베이직      SJ전자    매크로소프트        메론      성주모직
코드베이직   1.000000  0.091779 -0.048563 -0.067439 -0.052294
SJ전자    0.091779  1.000000 -0.001532  0.253515  0.092174
매크로소프트 -0.048563 -0.001532  1.000000 -0.055054 -0.039458
메론     -0.067439  0.253515 -0.055054  1.000000  0.172104
성주모직   -0.052294  0.092174 -0.039458  0.172104  1.000000

In [406]:
# TODO: 코드베이직과 가장 상관도가 높은 종목을 순서대로 두 개 표시한다.

## 정렬과 순위

In [323]:
Series(range(4), index=['d','a','b','c']).sort_index()

a    1
b    2
c    3
d    0
dtype: int64

In [328]:
frame = DataFrame(np.arange(8).reshape(2,4),
                 index=[u'다', u'가'],
                 columns=['d','a','b','c'])

In [329]:
frame.sort_index()

d  a  b  c
가  4  5  6  7
다  0  1  2  3

In [330]:
frame.sort_index(axis=1)

a  b  c  d
다  1  2  3  0
가  5  6  7  4

행열을 동시에 (사실은 순차적으로) 정렬

In [327]:
frame.sort_index().sort_index(axis=1)

a  b  c  d
가  5  6  7  4
다  1  2  3  0

In [331]:
frame.sort_index(axis=1, ascending=False)

d  c  b  a
다  0  3  2  1
가  4  7  6  5

In [334]:
Series(np.random.randint(10, size=5)).order()

1    0
4    1
3    3
0    8
2    8
dtype: int64

In [335]:
Series([4, np.nan, 7, np.nan, -3, 2]).order()

4    -3
5     2
0     4
2     7
1   NaN
3   NaN
dtype: float64

In [336]:
frame = DataFrame({'b': [4,-7,-3,2], 'a':[0,1,0,1]})
frame

a  b
0  0  4
1  1 -7
2  0 -3
3  1  2

In [337]:
frame.sort_index(by='b')

a  b
1  1 -7
2  0 -3
3  1  2
0  0  4

In [339]:
frame.sort_index(by=['b', 'a'])

a  b
1  1 -7
2  0 -3
3  1  2
0  0  4

In [349]:
sr = Series([7, -5, 7, 4, 2, 0, 4, 7])
sr

0    7
1   -5
2    7
3    4
4    2
5    0
6    4
7    7
dtype: int64

In [348]:
sr.rank()

0    7.0
1    1.0
2    7.0
3    4.5
4    3.0
5    2.0
6    4.5
7    7.0
dtype: float64

In [350]:
sr.rank(method='first')

0    6
1    1
2    7
3    4
4    3
5    2
6    5
7    8
dtype: float64

In [351]:
sr.rank(ascending=False, method='max')

0    3
1    8
2    3
3    5
4    6
5    7
6    5
7    3
dtype: float64

In [344]:
frame = DataFrame({'b': [4.3, 7, -3, 2],
                  'a': [0,1,0,1],
                  'c': [-2,5,8,-2.5]})
frame

a    b    c
0  0  4.3 -2.0
1  1  7.0  5.0
2  0 -3.0  8.0
3  1  2.0 -2.5

In [345]:
frame.rank(axis=1)

a  b  c
0  2  3  1
1  1  3  2
2  2  1  3
3  2  3  1

## 중복 색인

In [372]:
obj = Series(range(5), index=['a','a','b','b','c'])
obj

a    0
a    1
b    2
b    3
c    4
dtype: int64

In [373]:
obj.index.is_unique

False

In [374]:
obj['a']

a    0
a    1
dtype: int64

In [375]:
obj['c']

4

In [376]:
df = DataFrame(np.random.randn(4,3), index=['a','a','b','b'])
df

0         1         2
a -0.882303 -1.711796 -0.020002
a -1.530863  0.446040 -0.221814
b -0.101324  1.749098  1.102842
b -0.129260  0.232342  0.925445

In [378]:
df.index.is_unique

False

In [379]:
df.columns.is_unique

True

In [377]:
df.ix['b']

0         1         2
b -0.101324  1.749098  1.102842
b -0.129260  0.232342  0.925445

## 기술통계 계산과 요약

In [380]:
df = DataFrame([[1.4, np.nan],[7.1, -4.5],
               [np.nan, np.nan], [0.75, -1.3]],
              index=list('abcd'),
              columns=['one', 'two'])
df

one  two
a  1.40  NaN
b  7.10 -4.5
c   NaN  NaN
d  0.75 -1.3

In [381]:
df.sum()

one    9.25
two   -5.80
dtype: float64

In [382]:
df.sum(axis=1)

a    1.40
b    2.60
c     NaN
d   -0.55
dtype: float64

In [385]:
df.mean(axis=1)

a    1.400
b    1.300
c      NaN
d   -0.275
dtype: float64

In [383]:
df.mean(axis=1, skipna=False)

a      NaN
b    1.300
c      NaN
d   -0.275
dtype: float64

In [386]:
df.idxmax()

one    b
two    d
dtype: object

In [387]:
df.cumsum()

one  two
a  1.40  NaN
b  8.50 -4.5
c   NaN  NaN
d  9.25 -5.8

In [390]:
df.cumprod()

one   two
a  1.400   NaN
b  9.940 -4.50
c    NaN   NaN
d  7.455  5.85

In [388]:
df.describe()

one       two
count  3.000000  2.000000
mean   3.083333 -2.900000
std    3.493685  2.262742
min    0.750000 -4.500000
25%    1.075000 -3.700000
50%    1.400000 -2.900000
75%    4.250000 -2.100000
max    7.100000 -1.300000

In [389]:
obj = Series(['a','a', 'b', 'c']*4)
obj.describe()

count     16
unique     3
top        a
freq       8
dtype: object

## 상관관계와 공분산

In [391]:
web.get_data_yahoo('GOOG', '1/1/2000', '1/1/2010') # 버그?

Open    High     Low   Close  Volume
Date                                              
2004-08-19   49.96   51.98   47.93   50.12     NaN
2004-08-20   50.69   54.49   50.20   54.10     NaN
2004-08-23   55.32   56.68   54.47   54.65     NaN
2004-08-24   55.56   55.74   51.73   52.38     NaN
2004-08-25   52.43   53.95   51.89   52.95     NaN
2004-08-26   52.42   53.92   52.28   53.90     NaN
2004-08-27   54.00   54.26   52.79   53.02     NaN
2004-08-30   52.69   52.69   50.95   50.95     NaN
2004-08-31   51.11   51.80   51.03   51.13     NaN
2004-09-01   51.30   51.43   49.79   50.07     NaN
2004-09-02   49.50   51.13   49.42   50.70     NaN
2004-09-03   50.42   50.82   49.61   49.95     NaN
2004-09-07   50.45   50.95   49.76   50.74     NaN
2004-09-08   50.32   51.46   50.20   51.10     NaN
2004-09-09   51.20   51.30   50.45   51.10     NaN
2004-09-10   50.75   53.23   50.60   52.61     NaN
2004-09-13   53.26   54.15   53.18   53.70     NaN
2004-09-14   53.67   55.94   53.34   55.69     NaN
2004-09-15   55.22   57.06   55.04   55.94     NaN
2004-09-16   56.11   57.84   55.77   56.93     NaN
2004-09-17   57.15   58.69   56.72   58.69     NaN
2004-09-20   58.42   60.74   58.33   59.62     NaN
2004-09-21   60.04   60.15   58.70   58.86     NaN
2004-09-22   58.67   59.78   58.35   59.13     NaN
2004-09-23   59.36   61.25   58.45   60.35     NaN
2004-09-24   60.42   61.99   59.82   59.86     NaN
2004-09-27   59.72   60.38   58.84   59.07     NaN
2004-09-28   60.51   63.64   60.04   63.37     NaN
2004-09-29   63.20   67.44   63.05   65.47     NaN
2004-09-30   64.88   66.08   64.44   64.74     NaN
...            ...     ...     ...     ...     ...
2009-11-18  288.04  289.10  285.75  288.04     NaN
2009-11-19  286.60  286.71  284.72  286.21     NaN
2009-11-20  284.46  285.51  284.42  284.70     NaN
2009-11-23  287.96  293.01  287.64  290.88     NaN
2009-11-24  290.96  291.85  287.98  291.25     NaN
2009-11-25  292.91  293.24  291.05  292.58     NaN
2009-11-27  285.71  290.94  285.20  289.59     NaN
2009-11-30  290.02  291.54  288.27  291.21     NaN
2009-12-01  293.77  295.31  291.21  294.64     NaN
2009-12-02  295.19  296.21  292.82  293.46     NaN
2009-12-03  294.23  295.43  292.21  292.58     NaN
2009-12-04  296.21  297.12  289.30  292.21     NaN
2009-12-07  291.82  294.05  290.21  292.83     NaN
2009-12-08  291.46  295.03  290.71  293.23     NaN
2009-12-09  293.46  294.37  291.50  294.22     NaN
2009-12-10  294.92  297.06  294.91  295.45     NaN
2009-12-11  297.04  297.08  293.57  294.96     NaN
2009-12-14  297.32  298.36  296.01  297.57     NaN
2009-12-15  296.35  297.89  295.20  296.27     NaN
2009-12-16  299.16  299.88  298.02  298.58     NaN
2009-12-17  297.92  298.52  296.58  296.67     NaN
2009-12-18  297.72  299.17  297.20  297.91     NaN
2009-12-21  298.68  299.62  297.54  299.04     NaN
2009-12-22  300.37  300.45  299.13  300.26     NaN
2009-12-23  301.45  306.13  301.12  305.53     NaN
2009-12-24  306.16  309.45  305.83  308.93     NaN
2009-12-28  310.52  312.68  308.93  311.12     NaN
2009-12-29  312.06  312.11  308.84  309.39     NaN
2009-12-30  308.94  311.05  308.70  311.05     NaN
2009-12-31  312.06  312.39  309.68  309.68     NaN

[1353 rows x 5 columns]

In [394]:
web.get_data_google('GOOG', '1/1/2000', '1/1/2010') # Google Fiance로 변경

Open    High     Low   Close  Volume
Date                                              
2004-08-19   49.96   51.98   47.93   50.12     NaN
2004-08-20   50.69   54.49   50.20   54.10     NaN
2004-08-23   55.32   56.68   54.47   54.65     NaN
2004-08-24   55.56   55.74   51.73   52.38     NaN
2004-08-25   52.43   53.95   51.89   52.95     NaN
2004-08-26   52.42   53.92   52.28   53.90     NaN
2004-08-27   54.00   54.26   52.79   53.02     NaN
2004-08-30   52.69   52.69   50.95   50.95     NaN
2004-08-31   51.11   51.80   51.03   51.13     NaN
2004-09-01   51.30   51.43   49.79   50.07     NaN
2004-09-02   49.50   51.13   49.42   50.70     NaN
2004-09-03   50.42   50.82   49.61   49.95     NaN
2004-09-07   50.45   50.95   49.76   50.74     NaN
2004-09-08   50.32   51.46   50.20   51.10     NaN
2004-09-09   51.20   51.30   50.45   51.10     NaN
2004-09-10   50.75   53.23   50.60   52.61     NaN
2004-09-13   53.26   54.15   53.18   53.70     NaN
2004-09-14   53.67   55.94   53.34   55.69     NaN
2004-09-15   55.22   57.06   55.04   55.94     NaN
2004-09-16   56.11   57.84   55.77   56.93     NaN
2004-09-17   57.15   58.69   56.72   58.69     NaN
2004-09-20   58.42   60.74   58.33   59.62     NaN
2004-09-21   60.04   60.15   58.70   58.86     NaN
2004-09-22   58.67   59.78   58.35   59.13     NaN
2004-09-23   59.36   61.25   58.45   60.35     NaN
2004-09-24   60.42   61.99   59.82   59.86     NaN
2004-09-27   59.72   60.38   58.84   59.07     NaN
2004-09-28   60.51   63.64   60.04   63.37     NaN
2004-09-29   63.20   67.44   63.05   65.47     NaN
2004-09-30   64.88   66.08   64.44   64.74     NaN
...            ...     ...     ...     ...     ...
2009-11-18  288.04  289.10  285.75  288.04     NaN
2009-11-19  286.60  286.71  284.72  286.21     NaN
2009-11-20  284.46  285.51  284.42  284.70     NaN
2009-11-23  287.96  293.01  287.64  290.88     NaN
2009-11-24  290.96  291.85  287.98  291.25     NaN
2009-11-25  292.91  293.24  291.05  292.58     NaN
2009-11-27  285.71  290.94  285.20  289.59     NaN
2009-11-30  290.02  291.54  288.27  291.21     NaN
2009-12-01  293.77  295.31  291.21  294.64     NaN
2009-12-02  295.19  296.21  292.82  293.46     NaN
2009-12-03  294.23  295.43  292.21  292.58     NaN
2009-12-04  296.21  297.12  289.30  292.21     NaN
2009-12-07  291.82  294.05  290.21  292.83     NaN
2009-12-08  291.46  295.03  290.71  293.23     NaN
2009-12-09  293.46  294.37  291.50  294.22     NaN
2009-12-10  294.92  297.06  294.91  295.45     NaN
2009-12-11  297.04  297.08  293.57  294.96     NaN
2009-12-14  297.32  298.36  296.01  297.57     NaN
2009-12-15  296.35  297.89  295.20  296.27     NaN
2009-12-16  299.16  299.88  298.02  298.58     NaN
2009-12-17  297.92  298.52  296.58  296.67     NaN
2009-12-18  297.72  299.17  297.20  297.91     NaN
2009-12-21  298.68  299.62  297.54  299.04     NaN
2009-12-22  300.37  300.45  299.13  300.26     NaN
2009-12-23  301.45  306.13  301.12  305.53     NaN
2009-12-24  306.16  309.45  305.83  308.93     NaN
2009-12-28  310.52  312.68  308.93  311.12     NaN
2009-12-29  312.06  312.11  308.84  309.39     NaN
2009-12-30  308.94  311.05  308.70  311.05     NaN
2009-12-31  312.06  312.39  309.68  309.68     NaN

[1353 rows x 5 columns]

In [397]:
import pandas.io.data as web

all_data = {}
for ticker in ['AAPL', 'IBM', 'MSFT', 'GOOG']:
    all_data[ticker]=web.get_data_google(ticker, '1/1/2005', '1/1/2010')
    
price = DataFrame({tic: data['Close'] for tic, data in all_data.iteritems()})
volume = DataFrame({tic: data['Volume'] for tic, data in all_data.iteritems()})

In [398]:
returns = price.pct_change()
returns[-5:]

AAPL      GOOG       IBM      MSFT
Date                                              
2009-12-24  0.034292  0.011128  0.004385  0.002587
2009-12-28  0.012391  0.007089  0.013326  0.005484
2009-12-29 -0.011909 -0.005561 -0.003477  0.007058
2009-12-30  0.012052  0.005365  0.005461 -0.013699
2009-12-31 -0.004300 -0.004404 -0.012597 -0.015504

In [400]:
returns.AAPL.corr(returns.MSFT)

0.39730546887475782

In [295]:
returns.MSFT.cov(returns.IBM)

0.00017109113552596703

In [401]:
returns.corr()

AAPL      GOOG       IBM      MSFT
AAPL  1.000000  0.455271  0.443480  0.397305
GOOG  0.455271  1.000000  0.421986  0.486849
IBM   0.443480  0.421986  1.000000  0.558613
MSFT  0.397305  0.486849  0.558613  1.000000

In [402]:
returns.cov()

AAPL      GOOG       IBM      MSFT
AAPL  0.000912  0.000320  0.000208  0.000237
GOOG  0.000320  0.000540  0.000152  0.000223
IBM   0.000208  0.000152  0.000240  0.000171
MSFT  0.000237  0.000223  0.000171  0.000390

In [298]:
returns.corrwith(returns.IBM)

AAPL    0.496871
IBM     1.000000
MSFT    0.559671
dtype: float64

In [299]:
returns.corrwith(volume)

AAPL   -0.009899
IBM     0.031369
MSFT   -0.056570
dtype: float64

## 유일값, 값 세기, 멤버십

In [411]:
sr = Series(['c','a','d','a','a','b','b','c','c'])
sr

0    c
1    a
2    d
3    a
4    a
5    b
6    b
7    c
8    c
dtype: object

In [415]:
unique = sr.unique()
unique

array(['c', 'a', 'd', 'b'], dtype=object)

In [416]:
sr.value_counts()

c    3
a    3
b    2
d    1
dtype: int64

In [417]:
pd.value_counts(sr)

c    3
a    3
b    2
d    1
dtype: int64

In [419]:
mask = sr.isin(['b','c'])
mask

0     True
1    False
2    False
3    False
4    False
5     True
6     True
7     True
8     True
dtype: bool

In [420]:
sr[mask] # b 또는 c 인 데이터만 선택

0    c
5    b
6    b
7    c
8    c
dtype: object

In [421]:
data = DataFrame({'Qu1': [1,3,4,3,4],
                 'Qu2': [2,3,1,2,3],
                 'Qu3': [1,5,2,4,4]})
data

Qu1  Qu2  Qu3
0    1    2    1
1    3    3    5
2    4    1    2
3    3    2    4
4    4    3    4

In [422]:
result = data.apply(pd.value_counts).fillna(0)
result

Qu1  Qu2  Qu3
1    1    1    1
2    0    2    1
3    2    2    0
4    2    0    2
5    0    0    1

# 누락 데이터 처리

In [426]:
string_data = Series(['aadfasdba', 'artichoke', np.nan, 'avocado'])
string_data

0    aadfasdba
1    artichoke
2          NaN
3      avocado
dtype: object

In [430]:
string_data.isnull()

0    False
1    False
2     True
3    False
dtype: bool

In [431]:
string_data.isnull().any()

True

### 누락데이터 선별

In [435]:
from numpy import nan

In [436]:
data = Series([1, nan, 3.5, nan, 7])
data.dropna()

0    1.0
2    3.5
4    7.0
dtype: float64

In [434]:
# 불리언 색인으로 직접 선별
data[data.notnull()]

0    1.0
2    3.5
4    7.0
dtype: float64

In [445]:
data = DataFrame([[1., 6.5, 3.], [1., nan, nan],
                 [nan, nan, nan],[nan, 6.5, 3.]])
data

0    1   2
0   1  6.5   3
1   1  NaN NaN
2 NaN  NaN NaN
3 NaN  6.5   3

In [450]:
data.dropna?

In [439]:
data.dropna(how='all')

0    1   2
0   1  6.5   3
1   1  NaN NaN
3 NaN  6.5   3

In [441]:
data.dropna(axis=1, how='all')

0    1   2
0   1  6.5   3
1   1  NaN NaN
2 NaN  NaN NaN
3 NaN  6.5   3

In [447]:
df = DataFrame(np.random.randn(7,3))
df.ix[:4, 1] = nan
df.ix[:2, 2] = nan
df

0         1         2
0  0.625784       NaN       NaN
1 -0.143543       NaN       NaN
2 -0.628510       NaN       NaN
3  1.633658       NaN -1.548192
4 -0.302301       NaN -0.465363
5  0.964609  1.215261 -1.016997
6 -0.182277  0.490251 -0.028373

In [449]:
df.dropna(thresh=3)

0         1         2
5  0.964609  1.215261 -1.016997
6 -0.182277  0.490251 -0.028373

### 누락 값 채우기

In [451]:
df.fillna(0)

0         1         2
0  0.625784  0.000000  0.000000
1 -0.143543  0.000000  0.000000
2 -0.628510  0.000000  0.000000
3  1.633658  0.000000 -1.548192
4 -0.302301  0.000000 -0.465363
5  0.964609  1.215261 -1.016997
6 -0.182277  0.490251 -0.028373

In [452]:
df.fillna({1:0.5, 3:-1})

0         1         2
0  0.625784  0.500000       NaN
1 -0.143543  0.500000       NaN
2 -0.628510  0.500000       NaN
3  1.633658  0.500000 -1.548192
4 -0.302301  0.500000 -0.465363
5  0.964609  1.215261 -1.016997
6 -0.182277  0.490251 -0.028373

In [455]:
df is df.fillna(0) # [drop/fill]na()는 기본적으로 새로운 객체 반환

False

In [453]:
df.fillna(0, inplace=True)
df

0         1         2
0  0.625784  0.000000  0.000000
1 -0.143543  0.000000  0.000000
2 -0.628510  0.000000  0.000000
3  1.633658  0.000000 -1.548192
4 -0.302301  0.000000 -0.465363
5  0.964609  1.215261 -1.016997
6 -0.182277  0.490251 -0.028373

In [456]:
df = DataFrame(np.random.randn(6,3))
df.ix[2:,1]= nan
df.ix[4:, 2] = nan
df

0         1         2
0 -0.273364  0.721317 -0.000989
1 -0.036537  0.494105  0.266310
2 -1.863969       NaN  1.203902
3  0.494615       NaN  0.848738
4  0.175175       NaN       NaN
5 -0.004667       NaN       NaN

In [457]:
df.fillna(method='ffill')

0         1         2
0 -0.273364  0.721317 -0.000989
1 -0.036537  0.494105  0.266310
2 -1.863969  0.494105  1.203902
3  0.494615  0.494105  0.848738
4  0.175175  0.494105  0.848738
5 -0.004667  0.494105  0.848738

In [458]:
df.fillna(method='ffill', limit=2)

0         1         2
0 -0.273364  0.721317 -0.000989
1 -0.036537  0.494105  0.266310
2 -1.863969  0.494105  1.203902
3  0.494615  0.494105  0.848738
4  0.175175       NaN  0.848738
5 -0.004667       NaN  0.848738

In [459]:
data = Series([1., nan, 3.5, nan, 7])
data.fillna(data.mean())

0    1.000000
1    3.833333
2    3.500000
3    3.833333
4    7.000000
dtype: float64

# 도전과제

공공데이터 활용

다음 코드저장소에서 제주도 관광객 유동인구 데이터를 다운로드 받는다.
https://github.com/lseongjoo/py_data_go_kr

1. 각 데이터 파일을 DataFrame으로 읽어들인다. (p.222-227)
2. 각 DataFrame의 STD_YM 열 제목을 '기준연월'로 변경 (p.170-172)
3. FLOW_ ...로 시작하는 컬럼 중, 값이 150 이상인 것만 선택 (p.180-183, 문자열 처리 관련: p.291-298)
4. 모든 DataFrame을 합쳐서 단일한 DataFrame으로 생성 (p.253-269)
5. 해당 셀의 유동인구 평균을 구하기 (p.195-198)

In [169]:
path='py_data_go_kr/data/jeju/'

In [170]:
path += 'cell_300m_201406/'
path += '3-2-1-2. 시간대별_관광객_300그리드단위_유동인구_비율포함_201406.txt'

In [171]:
open(path)

<open file 'py_data_go_kr/data/jeju/cell_300m_201406/3-2-1-2. \xe1\x84\x89\xe1\x85\xb5\xe1\x84\x80\xe1\x85\xa1\xe1\x86\xab\xe1\x84\x83\xe1\x85\xa2\xe1\x84\x87\xe1\x85\xa7\xe1\x86\xaf_\xe1\x84\x80\xe1\x85\xaa\xe1\x86\xab\xe1\x84\x80\xe1\x85\xaa\xe1\x86\xbc\xe1\x84\x80\xe1\x85\xa2\xe1\x86\xa8_300\xe1\x84\x80\xe1\x85\xb3\xe1\x84\x85\xe1\x85\xb5\xe1\x84\x83\xe1\x85\xb3\xe1\x84\x83\xe1\x85\xa1\xe1\x86\xab\xe1\x84\x8b\xe1\x85\xb1_\xe1\x84\x8b\xe1\x85\xb2\xe1\x84\x83\xe1\x85\xa9\xe1\x86\xbc\xe1\x84\x8b\xe1\x85\xb5\xe1\x86\xab\xe1\x84\x80\xe1\x85\xae_\xe1\x84\x87\xe1\x85\xb5\xe1\x84\x8b\xe1\x85\xb2\xe1\x86\xaf\xe1\x84\x91\xe1\x85\xa9\xe1\x84\x92\xe1\x85\xa1\xe1\x86\xb7_201406.txt', mode 'r' at 0x106afaae0>

In [185]:
df = pd.read_csv(path, sep='|')

In [198]:
id(df.rename(columns={'STD_YM': u'기준연월'}))

4409061968

In [199]:
id(df)

4409219984

STD_YM      ID_300  FLOW_POP_CNT_00TMST  FLOW_POP_CNT_01TMST  \
0      201406  1000031709           312.947980           263.972057   
1      201406  1000030747                  NaN                  NaN   
2      201406  1000031707           229.957330           194.422903   
3      201406  1000031388                  NaN                  NaN   
4      201406  1000031708           384.201369           320.784622   
5      201406  1000030741                  NaN                  NaN   
6      201406  1000032349           203.903159           164.497839   
7      201406  1000032027           236.215989           193.343749   
8      201406  1000031390                  NaN                  NaN   
9      201406  1000029783                  NaN                  NaN   
10     201406  1000031384                  NaN                  NaN   
11     201406  1000032028           290.680033           239.531440   
12     201406  1000031706           198.493194           167.362223   
13     201406  1000031705                  NaN                  NaN   
14     201406  1000031387                  NaN                  NaN   
15     201406  1000031389                  NaN                  NaN   
16     201406  1000031067                  NaN                  NaN   
17     201406  1000032026                  NaN                  NaN   
18     201406  1000031064                  NaN                  NaN   
19     201406  1000031386                  NaN                  NaN   
20     201406  1000031068                  NaN                  NaN   
21     201406  1000031062                  NaN                  NaN   
22     201406  1000030426                  NaN                  NaN   
23     201406  1000032668                  NaN                  NaN   
24     201406  1000030740                  NaN                  NaN   
25     201406  1000030748                  NaN                  NaN   
26     201406  1000030743                  NaN                  NaN   
27     201406  1000029776                  NaN                  NaN   
28     201406  1000031385                  NaN                  NaN   
29     201406  1000032029           330.171191           270.769124   
...       ...         ...                  ...                  ...   
10080  201406  1000044802                  NaN                  NaN   
10081  201406  1000028297                  NaN                  NaN   
10082  201406  1000041147                  NaN                  NaN   
10083  201406  1000051371                  NaN                  NaN   
10084  201406  1000050408                  NaN                  NaN   
10085  201406  1000034250                  NaN                  NaN   
10086  201406  1000062137                  NaN                  NaN   
10087  201406  1000048195                  NaN                  NaN   
10088  201406  1000042731                  NaN                  NaN   
10089  201406  1000041773                  NaN                  NaN   
10090  201406  1000036177                  NaN                  NaN   
10091  201406  1000033936                  NaN                  NaN   
10092  201406  1000057351                  NaN                  NaN   
10093  201406  1000051692                  NaN                  NaN   
10094  201406  1000058706                  NaN                  NaN   
10095  201406  1000033619                  NaN                  NaN   
10096  201406  1000034248                  NaN                  NaN   
10097  201406  1000033927                  NaN                  NaN   
10098  201406  1000042411                  NaN                  NaN   
10099  201406  1000041466                  NaN                  NaN   
10100  201406  1000060237                  NaN                  NaN   
10101  201406  1000027975                  NaN                  NaN   
10102  201406  1000052670                  NaN                  NaN   
10103  201406  1000060517                  NaN                  NaN   
10104  201406  1000034898      

In [201]:
text = 'abcdefg'

In [202]:
'f' in text

True

In [203]:
'z' in text

False

In [204]:
column_title = 'FLOW_ADBD_@#SD'

In [205]:
'FLOW' in column_title

True

In [206]:
column_title.startswith('FLOW')

True

In [207]:
text.startswith('FLOW')

False

In [211]:
df.ix[:, 2:26]

FLOW_POP_CNT_00TMST  FLOW_POP_CNT_01TMST  FLOW_POP_CNT_02TMST  \
0               312.947980           263.972057           227.917759   
1                18.193314            15.509624            13.867847   
2               229.957330           194.422903           171.347177   
3                94.882052            82.826656            72.077200   
4               384.201369           320.784622           278.623635   
5                 8.431997             6.973419             6.201712   
6               203.903159           164.497839           136.066924   
7               236.215989           193.343749           166.153644   
8                51.910113            44.783555            40.537079   
9                37.495007            29.403356            24.888177   
10               53.102493            44.402127            39.036083   
11              290.680033           239.531440           205.811635   
12              198.493194           167.362223           146.138694   
13               76.138139            62.224367            54.357331   
14               90.234275            78.440401            69.464986   
15               86.399788            75.414361            67.676082   
16               16.192821            14.774156            13.421957   
17              124.639615            96.210871            81.740989   
18               16.610321            14.390724            12.804391   
19              100.817733            85.175177            74.586041   
20               10.974405             9.711384             8.824614   
21               15.956015            13.022493            11.428144   
22                2.560444             2.149290             1.917661   
23               92.995119            73.742357            63.657060   
24               11.784689             9.839640             8.848829   
25                6.846833             5.638254             5.024362   
26                2.633757             2.260846             2.022704   
27                0.914158             0.742983             0.639276   
28               54.428974            44.782052            38.575497   
29              330.171191           270.769124           230.724194   
...                    ...                  ...                  ...   
10080             0.044923             0.041917             0.040581   
10081             0.042919             0.036573             0.035571   
10082             0.046426             0.040748             0.040581   
10083             0.089846             0.088343             0.089345   
10084             0.052104             0.050100             0.049933   
10085             0.129425             0.142618             0.140280   
10086             0.058951             0.052271             0.050601   
10087             0.087675             0.090514             0.090848   
10088             0.064128             0.066466             0.063293   
10089             0.088009             0.090848             0.091182   
10090             0.057281             0.050267             0.048931   
10091             0.062959             0.055778             0.053941   
10092             0.041583             0.041082             0.041416   
10093             0.058617             0.055778             0.055611   
10094             0.000167             0.000000             0.000000   
10095             0.079993             0.071977             0.069806   
10096             0.082164             0.089011             0.087508   
10097             0.106379             0.115397             0.112725   
10098             0.055611             0.058116             0.054943   
10099             0.034736             0.029726             0.029392   
10100             0.025050             0.023881             0.023380   
10101             0.068971             0.058450             0.058283   
10102             0.064796             0.062792             0.061456   
10103             0.018036             0.016366

## 계층 색인

In [461]:
data = Series(np.random.randn(10),
             index=[['a','a','a', 'b','b','b','c','c','d','d'],
                   [1,2,3,1,2,3,1,2,2,3]])
data

a  1    0.729613
   2    0.362837
   3    1.388591
b  1   -0.825556
   2    1.581476
   3   -2.297921
c  1    0.886037
   2    0.504794
d  2   -1.898794
   3    0.127336
dtype: float64

In [215]:
data.index

MultiIndex(levels=[[u'a', u'b', u'c', u'd'], [1, 2, 3]],
           labels=[[0, 0, 0, 1, 1, 1, 2, 2, 3, 3], [0, 1, 2, 0, 1, 2, 0, 1, 1, 2]])

In [214]:
type(data.index)

pandas.core.index.MultiIndex

In [216]:
data['b']

1    0.442951
2   -1.133848
3   -0.625826
dtype: float64

In [217]:
type(data['b'])

pandas.core.series.Series

In [218]:
data['b':'c']

b  1    0.442951
   2   -1.133848
   3   -0.625826
c  1   -2.355475
   2    0.176855
dtype: float64

In [219]:
data.ix[['b','d']]

b  1    0.442951
   2   -1.133848
   3   -0.625826
d  2    0.714299
   3   -0.341819
dtype: float64

In [223]:
data[:,2] # 첫 번째 인덱스 하위의 2 인덱스 출력

a   -0.176001
b   -0.625826
d   -0.341819
dtype: float64

In [462]:
data.unstack()

1         2         3
a  0.729613  0.362837  1.388591
b -0.825556  1.581476 -2.297921
c  0.886037  0.504794       NaN
d       NaN -1.898794  0.127336

In [228]:
type(data.unstack())

pandas.core.frame.DataFrame

In [229]:
data.unstack().stack() # 원래대로 돌아온다. 아무것도 안 한 것과 마찬가지.

a  1    0.263428
   2    0.008796
   3   -0.176001
b  1    0.442951
   2   -1.133848
   3   -0.625826
c  1   -2.355475
   2    0.176855
d  2    0.714299
   3   -0.341819
dtype: float64

In [464]:
frame = DataFrame(np.arange(12).reshape((4,3)),
                 index=[['a','a','b','b'], [1,2,1,2]],
                 columns=[['Ohio','Ohio', 'Colorado'],
                         ['Green','Red','Green']])
frame

Ohio     Colorado
    Green Red    Green
a 1     0   1        2
  2     3   4        5
b 1     6   7        8
  2     9  10       11

In [465]:
frame.index.names = ['key1', 'key2']
frame.columns.names = ['state', 'color']
frame

state      Ohio     Colorado
color     Green Red    Green
key1 key2                   
a    1        0   1        2
     2        3   4        5
b    1        6   7        8
     2        9  10       11

In [469]:
frame['Ohio']

color      Green  Red
key1 key2            
a    1         0    1
     2         3    4
b    1         6    7
     2         9   10

In [237]:
frame.swaplevel('key1','key2')

state      Ohio     Colorado
color     Green Red    Green
key2 key1                   
1    a        0   1        2
2    a        3   4        5
1    b        6   7        8
2    b        9  10       11

In [242]:
frame.sortlevel(1)

state      Ohio     Colorado
color     Green Red    Green
key1 key2                   
a    1        0   1        2
b    1        6   7        8
a    2        3   4        5
b    2        9  10       11

In [470]:
frame.swaplevel(0,1).sortlevel(0)

state      Ohio     Colorado
color     Green Red    Green
key2 key1                   
1    a        0   1        2
     b        6   7        8
2    a        3   4        5
     b        9  10       11

In [243]:
frame.sum()

state     color
Ohio      Green    18
          Red      22
Colorado  Green    26
dtype: int64

In [471]:
frame.sum(level='key2')

state  Ohio     Colorado
color Green Red    Green
key2                    
1         6   8       10
2        12  14       16

In [248]:
frame.sum(level='color', axis=1)

color      Green  Red
key1 key2            
a    1         2    1
     2         8    4
b    1        14    7
     2        20   10

In [472]:
frame = DataFrame({'a': range(7),
                  'b': range(7,0, -1),
                  'c': ['one', 'one', 'one', 'two', 'two', 'two', 'two'],
                  'd':[0,1,2,0,1,2,3]})
frame

a  b    c  d
0  0  7  one  0
1  1  6  one  1
2  2  5  one  2
3  3  4  two  0
4  4  3  two  1
5  5  2  two  2
6  6  1  two  3

In [473]:
frame2 = frame.set_index(['c','d'])
frame2

a  b
c   d      
one 0  0  7
    1  1  6
    2  2  5
two 0  3  4
    1  4  3
    2  5  2
    3  6  1

In [476]:
frame.set_index(['c','d'], drop=False)

a  b    c  d
c   d              
one 0  0  7  one  0
    1  1  6  one  1
    2  2  5  one  2
two 0  3  4  two  0
    1  4  3  two  1
    2  5  2  two  2
    3  6  1  two  3

In [252]:
frame2.reset_index()

c  d  a  b
0  one  0  0  7
1  one  1  1  6
2  one  2  2  5
3  two  0  3  4
4  two  1  4  3
5  two  2  5  2
6  two  3  6  1

In [276]:
df1

one  two
a    0    1
b    2    3
c    4    5

In [277]:
df2

three  four
a      0     1
c      2     3

In [280]:
pd.concat([df1, df2], axis=1)

one  two  three  four
a    0    1      0     1
b    2    3    NaN   NaN
c    4    5      2     3

In [281]:
a = Series([np.nan, 2.5, np.nan, 3.5, 4.5, np.nan],
          index=['f', 'e', 'd', 'c', 'b', 'a'])
b = Series(np.arange(len(a), dtype=np.float64),
          index=['f', 'e', 'd', 'c', 'b', 'a'])

In [282]:
a

f    NaN
e    2.5
d    NaN
c    3.5
b    4.5
a    NaN
dtype: float64

In [286]:
b

f     0
e     1
d     2
c     3
b     4
a   NaN
dtype: float64

In [285]:
b[-1]=np.nan

In [289]:
# 각 인덱스에서, a의 값이 없는 경우는 b의 해당 인덱스의 값을 가져온다
np.where(pd.isnull(a), b, a)

array([ 0. ,  2.5,  2. ,  3.5,  4.5,  nan])

In [288]:
pd.isnull(a)

f     True
e    False
d     True
c    False
b    False
a     True
dtype: bool